log4j-core/src/main/java/org/apache/logging/log4j/core/pattern/NamedDatePattern.java

```diff
@@ -0,0 +1,114 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.pattern;
+
+import static org.apache.logging.log4j.core.util.internal.instant.InstantPatternFormatter.LEGACY_FORMATTERS_ENABLED;
+
+import org.apache.logging.log4j.core.util.internal.instant.InstantPatternFormatter;
+
+/**
+ * Represents named date & time patterns for formatting log timestamps.
+ *
+ * @see InstantPatternFormatter#LEGACY_FORMATTERS_ENABLED
+ * @see DatePatternConverter
+ * @since 2.26.0
+ */
+public enum NamedDatePattern {
+
+    // If legacy formatters are enabled, we need to produce output aimed for `FixedDateFormat` and `FastDateFormat`.
+    // Otherwise, we need to produce output aimed for `DateTimeFormatter`.
+    // In conclusion, we need to check if legacy formatters enabled and apply following transformations.
+    //
+    //                               | Microseconds | Nanoseconds | Time-zone
+    // ------------------------------+--------------+-------------+-----------
+    // Legacy formatter directive    | nnnnnn       | nnnnnnnnn   | X, XX, XXX
+    // `DateTimeFormatter` directive | SSSSSS       | SSSSSSSSS   | x, xx, xxx
+    //
+    // Enabling legacy formatters mean that user requests the pattern to be formatted using deprecated
+    // `FixedDateFormat` and `FastDateFormat`.
+    // These two have, let's not say _bogus_, but an _interesting_ way of handling certain pattern directives:
+    //
+    // - They say they adhere to `SimpleDateFormat` specification, but use `n` directive.
+    //   `n` is neither defined by `SimpleDateFormat`, nor `SimpleDateFormat` supports sub-millisecond precisions.
+    //   `n` is probably manually introduced by Log4j to support sub-millisecond precisions.
+    //
+    // - `n` denotes nano-of-second for `DateTimeFormatter`.
+    //   In Java 17, `n` and `N` (nano-of-day) always output nanosecond precision.
+    //   This is independent of how many times they occur consequently.
+    //   Yet legacy formatters use repeated `n` to denote sub-milliseconds precision of certain length.
+    //   This doesn't work for `DateTimeFormatter`, which needs
+    //
+    //   - `SSSSSS` for 6-digit microsecond precision
+    //   - `SSSSSSSSS` for 9-digit nanosecond precision
+    //
+    // - Legacy formatters use `X`, `XX,` and `XXX` to choose between `+00`, `+0000`, or `+00:00`.
+    //   This is the correct behaviour for `SimpleDateFormat`.
+    //   Though `X` in `DateTimeFormatter` produces `Z` for zero-offset.
+    //   To avoid the `Z` output, one needs to use `x` with `DateTimeFormatter`.
```

Comment:
Of course this comment can be removed once we remove legacy patterns from this class.